### DISCUSSION TITLE
Specifying Minimum and Maximum X location constraint

### DISCUSSION THREAD

**Discussion by ardao13** [su2code/SU2#2162] [General]
I am trying to optimize a protuberance shape. 

Instead of specifying a chord length constraint, I want to add minimum and maximum x locations for the protuberance shape as constraints. 

Can anyone help ?

**Comment by TobiKattmann** [comment]
As a start it would be helpful to attach some pictures, and/or a sample case such the problem is better understandable. I guess that you work with FFD boxes and not with Hicks Henne bump functions. In that case I do believe it is probably easier to use a third party optimizer FADO https://github.com/pcarruscag/fado and introduce the constraint manually there. Please understand that FADO cannot magically do that, but you need to find a way to evaluate Minimum and Maximum X location yourself via some function and maybe even how to resolve that instead of rejecting the violating design (but the second part is probably not the first step I guess)

**Reply by ardao13** [reply]
Dear Tobi, thank you for your response. As you suggest, first let me provide a clearer overview of my problem with the figure below.

![Screenshot from 2023-12-09 16-25-47](https://github.com/su2code/SU2/assets/73038236/a6e1585a-217c-4f83-a2f4-96195b820cf1)


Currently, I'm utilizing the shape_optimization.py with the discrete adjoint method for optimization, and it's progressing well. However, I now want to incorporate space constraints into the process. Following your suggestion, I'm in the process of learning FADO through tutorials of Species_Transport Design

My specific question is whether I should create an entire script for the optimization process similar to the one found at Species_Transport tutorials and include the constraints there, or can I use FADO specifically to enhance the features of the shape optimization.py?

Which way should I focus?

Thanks again for your guidance!

**Reply by bigfooted** [reply]
Hi, there might be much better ways of doing this, but one thing you could do for FFD boxes is simply use the `bounds` option in the scipy optimizer:
https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html
https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.Bounds.html#scipy.optimize.Bounds
You just use `bounds=mybounds()` in the optimizer command and you create a file mybounds.py that returns the constraints for each of the nodes in the FFD box, e.g. 
```
bounds=(
(p1_x_lowerbound,p1_x_upperbound),
(p2_x_lowerbound,p2_x_upperbound),
...,
(p1_y_lowerbound,p1_y_upperbound),
(p2_y_lowerbound,p2_y_upperbound),
...)
```

**Reply by ardao13** [reply]
Dear @bigfooted,

Thanks for pointing me in the right direction. I've successfully implemented the bounds using **scipy.optimize.fmin_slsqp** in the **shape_optimization.py** file, and it works!  Now, I don't need to create a script with FADO. I made a simple modification to the shape_optimization.py file, allowing it to read bounds as a **list** instead of **float** from the configuration file. 



`OPT_BOUND_LOWER=-0.0191, -98.9, -100.4, -0.01919 ...continues like that and they match with DV_VALUE`


Modification of shape_optimization.py:


    # modified------------------------------------------------------        
    xb_up= [float(_)/float(relax_factor) for _ in config.OPT_BOUND_UPPER.split(",")]
    xb_low= [float(_)/float(relax_factor) for _ in config.OPT_BOUND_LOWER.split(",")]
    xb = list(zip(xb_low, xb_up))  # design bounds
    #------------------------------------------------------------------------------------

**Reply by bigfooted** [reply]
Great! I still encourage you to use FADO though :-)
